### PR TITLE
Output DWARF informations for the enum declarations

### PR DESCRIPTION
**Pull request by LemonBoy**
As noted by @mathias-lang-sociomantic in [this ticket](https://issues.dlang.org/show_bug.cgi?id=15630) `gdb` wouldn't show the value of the enum-typed parameter.
Further investigation showed that this was never implemented or has been thrown away at some point so I took the chance and implemented it and also extended the functionality a bit: it now works with every enum with an integral base type, both signed and unsigned, ranging from {u,}byte to {u,}long.

It hasn't been tested on x86, only x64.

Some technical infos: the `SENforward` flag is now ignored since it was only set once in the whole codebase (in `type_enum` and never cleared) and is not needed for our purpose (I can remove the whole `#if 0` block and/or the flag itself if needed). We now emit a `DW_AT_type` for the enumeration type like other cooler compilers do (I've cross-checked this with `LDC`, we're pretty cool).

### CODE REVIEW

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Check should be for isintegral and ty != Tbool.

**Comment by ibuclaw** [comment]
> cross-checked this with LDC

I'd cross check with gdc as it's the only compiler that has always had working debug generation. ;)

**Comment by ibuclaw** [comment]
Probably add a comment that DW_AT_TYPE represents the underlying enum type.

By the way, are named enums marked with the class tag?  As they are equivalent to C++ enum : class.

**Comment by LemonBoy** [comment]
> Check should be for isintegral and ty != Tbool.

I prefer not to alias the enum to its base type, if `enum Foo : bool` is the type I've declared in my code then I'd expect to see `Foo` and not `bool`.

> I'd cross check with gdc as it's the only compiler that has always had working debug generation. ;)

Didn't have a copy at hand and the ArchLinux package for it seems to have some problems during the linking phase. I have checked the results against GDC too :)

> Probably add a comment that DW_AT_TYPE represents the underlying enum type.

Yeah, but that's the only type that comes into play so I'm not sure it adds anything.

> By the way, are named enums marked with the class tag? As they are equivalent to C++ enum : class.

Good question, I'll add [it](http://www.dwarfstd.org/ShowIssue.php?issue=090303.1) as it makes the output much neater :)

**Comment by ibuclaw** [comment]
> Good question, I'll add it as it makes the output much neater :)

Yeah, it should be:
```
(gdb) print foo
$1 = (Type.MEMBER1 | Type.MEMBER2)
```
vs.
```
$1 = (MEMBER1 | MEMBER2)
```

**Comment by ibuclaw** [comment]
> I prefer not to alias the enum to its base type, if enum Foo : bool is the type I've declared in my code then I'd expect to see Foo and not bool.

The thing is, bool is *not* a valid enum type.  GCC's `dwarf2out.c` for sure ICE's if it finds this,  so I'd expect GDB's `dwarf2read.c` to do the same thing, and if it doesn't, then it probably should.

**Comment by ibuclaw** [comment]
> I prefer not to alias the enum to its base type, if enum Foo : bool is the type I've declared in my code then I'd expect to see Foo and not bool.

For non-integral types (bool including), you could use a typedef for them.

**Comment by LemonBoy** [comment]
> The thing is, bool is not a valid enum type. gcc's dwarf2out.c for sure ICE's if it finds this, so I'd expect gdb's dwarf2read.c to do the same thing, and if it doesn't, then it probably should.

Well, gdb has been able to parse and use such a type just fine, maybe they've recently fixed this.

> For non-integral types (bool including), you could use a typedef for them.

Neither gdc nor ldc seem to produce any tag in such cases so why bother :)
Actually gdc wins since ldc outputs the `enum` tags for `float` and `double` typed enums with not-so-good results (`f=3.14` becomes `f=4617473143948558991` :) and truncates the values to their integer part only.

---

I can't figure what the hell is going on with the autotester but it seems to be a mix of `targ_llong` being used and a weird-ass error in druntime's `hashOf`...

**Comment by ibuclaw** [comment]
> Neither gdc nor ldc seem to produce any tag in such cases so why bother :)

Yeah, I have not released any such patch in gdc yet.  I was just toying the idea as a possible alternative to give you the strong type name - but without the named values.

> Actually gdc wins since ldc outputs the enum tags for float and double typed enums with not-so-good results 

That llvm allows that sounds like a bug in their compiler...

**Comment by ibuclaw** [comment]
> Well, gdb has been able to parse and use such a type just fine, maybe they've recently fixed this.

Still doesn't make it correct.

**Comment by LemonBoy** [comment]
> Still doesn't make it correct.

The DWARF4 manual doesn't put any restriction on the type of the enumeration, the only problem here is that `gdb` doesn't like any non-integral type and it just keeps printing the values as integers.

**Comment by MartinNowak** [comment]
In any case please add a gdb test, see https://github.com/dlang/dmd/blob/2f19beaae1190102c08c9fa6a4b0a2f6c0813425/test/runnable/gdb1.d for an example.

**Comment by LemonBoy** [comment]
It seems that druntime doesn't quite like this modification but couldn't figure out why yet, once the problem is solved I'll add a gdbtest too.

**Comment by ibuclaw** [comment]
> The DWARF4 manual doesn't put any restriction on the type of the enumeration

That's probably because any mention of enum implies what is described as an enum for C/C++. ;-) 

If anything else works, I would consider it an accident, and it may break in future.

**Comment by JinShil** [comment]
@ibuclaw Does this PR have a future?  If so, what needs to be done?

**Comment by ibuclaw** [comment]
I'm waiting for a response from the gdb maintainers regarding `enum : bool`.  If they say no, then my change request stands.

If the condition is fixed up as per my request, then this patch is fine.

**Comment by JinShil** [comment]
Attempted revival at #7388

**Comment by JinShil** [comment]
#7388 Currently isn't successful, but it has a much better chance than this PR.  So closing.  Again, thanks to @LemonBoy for the initial work.

**Pull request closed without merging**
